src/python/pants/backend/go/pkg.py

```diff
@@ -0,0 +1,256 @@
+        if not resolved_go_module.import_path:
+            raise ValueError(
+                f"Unable to infer import path for the `go_package` at address {request.address} "
+                f"because the owning go_module at address {resolved_go_module.address} "
+                "does not have an import path defined nor could one be inferred."
+            )
+        import_path = f"{resolved_go_module.import_path}{spec_subpath}"
+
+    sources = await Get(
+        SourceFiles,
+        SourceFilesRequest(
+            [
+                target.get(GoPackageSources),
+                resolved_go_module.target.get(GoModuleSources),
+            ]
+        ),
+    )
+
+    # Note: The `go` tool requires GOPATH to be an absolute path which can only be resolved from within the
+    # execution sandbox. Thus, this code uses a bash script to be able to resolve that path.
+    # TODO: Merge all duplicate versions of this script into a single script and add an invoke rule that will
+    # insert the desired `go` command into the boilerplate portions.
+    analyze_script_digest = await Get(
+        Digest,
+        CreateDigest(
+            [
+                FileContent(
+                    "analyze.sh",
+                    textwrap.dedent(
+                        f"""\
+                export GOROOT="$(/bin/pwd)/go"
+                export GOPATH="$(/bin/pwd)/gopath"
+                export GOCACHE="$(/bin/pwd)/cache"
+                /bin/mkdir -p "$GOPATH" "$GOCACHE"
+                cd {resolved_go_module.address.spec_path}
+                exec "${{GOROOT}}/bin/go" list -json ./{spec_subpath}
+                """
+                    ).encode("utf-8"),
+                )
+            ]
+        ),
+    )
+
+    input_root_digest = await Get(
+        Digest,
+        MergeDigests([sources.snapshot.digest, downloaded_goroot.digest, analyze_script_digest]),
+    )
+
+    process = Process(
+        argv=[bash.path, "./analyze.sh"],
+        input_digest=input_root_digest,
+        description="Resolve go_package metadata.",
+        level=LogLevel.DEBUG,
+    )
+
+    result = await Get(ProcessResult, Process, process)
+
+    metadata = json.loads(result.stdout)
+
+    # TODO: Raise an exception on errors. They are only emitted as warnings for now because the `go` tool is
+    # flagging missing first-party code as a dependency error. But we want dependency inference and won't know
+    # what the dependency actually is unless we first resolve the package with that dependency. So circular
+    # reasoning. We may need to hydrate the sources for all go_package targets that share a `go_module`.
+    if metadata.get("Incomplete"):
+        error_dict = metadata.get("Error", {})
+        if error_dict:
+            error_str = error_to_string(error_dict)
+            _logger.warning(
+                f"Error while resolving Go package at address {request.address}: {error_str}"
+            )
+        # TODO: Check DepsErrors key as well.
```

Comment:
I need to see what shows up in practice before I can do this. This will likely be made into an error during my cleanup of this code next week. This is a development-only intermediate form of this code.